bin/storj-contract-renew.js

```diff
@@ -0,0 +1,180 @@
+}
+
+/**
+ * Inspects the contract to ensure it's still valid and in use
+ * If the end time is within 24 hours, renew it for another 3 months
+ * @function
+ * @param {object} shard
+ */
+function handleCursorData(shard) {
+  let needsRenewal = [];
+
+  cursor.pause();
+
+  for (let contractInfo of shard.contracts) {
+    let { contract: contractObj, nodeID } = contractInfo;
+    let contract = new storj.Contract(contractObj);
+
+    needsRenewal.push([nodeID, contract]);
+  }
+
+  let renewalContracts = needsRenewal.map(([nodeId, contract]) => {
+    contract.set('store_begin', NOW);
+    contract.set('store_end', NOW + ms('90d'));
+
+    return [nodeId, contract];
+  });
+
+  async.map(renewalContracts, lookupFarmer, maybeRenewContracts);
+}
+
+function lookupFarmer([nodeId, contract], next) {
+  storage.models.Contact.findOne({ _id: nodeId }, (err, contact) => {
+    if (err || !contact) {
+      counter.errored++;
+      return next(null, null);
+    }
+
+    next(null , [storj.Contact(contact.toObject()), contract]);
+  });
+}
+
+function maybeRenewContracts(err, results) {
+  let canBeRenewed = results.filter((result) => !!result);
+
+  async.eachLimit(canBeRenewed, 6, checkIfNeedsRenew,
+                  () => cursor.resume());
+}
+
+function finishProcessingContract(done) {
+  counter.processed++;
+  done();
+}
+
+function checkIfNeedsRenew([contact, contract], done) {
+  let shardHash = contract.get('data_hash');
+
+  async.waterfall(
+    [
+      (next) => getPointerObjects(shardHash, next),
+      (pointers, next) => getAssociatedFrames(pointers, next),
+      (frames, next) => getParentBucketEntries(frames, next),
+      (next) => renewContract([contact, contract], next)
+    ],
+    () => finishProcessingContract(done)
+  );
+}
+
+function getPointerObjects(shardHash, next) {
+  storage.models.Pointer.find({ hash: shardHash }, (err, pointers) => {
+    next(err || pointers.length === 0, pointers);
+  });
+}
+
+function getAssociatedFrames(pointers, next) {
+  storage.models.Frame.find({
+    shards: {
+      $in: pointers.map((pointer) => pointer._id)
+    }
+  }, (err, frames) => {
+    next(err || frames.length === 0, frames);
+  });
+}
+
+function getParentBucketEntries(frames, next) {
+  storage.models.BucketEntry.find({
+    frame: {
+      $in: frames.map((frame) => frame._id)
+    }
+  }, (err, entries) => {
+    next(err || entries.length === 0);
+  });
+}
+
+function updateContractRecord(contact, contract, next) {
+  storage.models.Shard.findOne({
+    hash: contract.get('data_hash')
+  }, function(err, shard) {
+    if (err) {
+      return next(err);
+    }
+
+    for (let contract of shard.contracts) {
+      if (contract.nodeID !== contact.nodeID) {
+        continue;
+      }
+
+      contract.contract = contract.toObject();
+    }
+
+    shard.save(next);
+  });
+}
+
+function renewContract([contact, contract], next) {
+  network.renewContract(contract, contact, (err) => {
+    if (err) {
+      counter.errored++;
```

Comment:
Updated to aggregate errors in a list with contact, contract, and error message to be included in the result output.